packages/react-server/src/ReactFlightServer.js

```diff
@@ -648,12 +662,81 @@ function createTask(
   abortSet: Set<Task>,
 ): Task {
   const id = request.nextChunkId++;
+  if (typeof model === 'object' && model !== null) {
+    // Register this model as having the ID we're about to write.
+    request.writtenObjects.set(model, id);
```

Comment:
Notably this doesn't tag things that are returned from a Server Component at this ID as having the same ID as the Element itself. So it might outline the return value again instead of reusing the same ID.

However, I'm not sure Server Components will be reusable with some other changes coming anyway.